src/lib/operators/join_nested_loop.cpp

```diff
@@ -0,0 +1,264 @@
+      break;
+    case ScanType::OpGreaterThanEquals:
+      _comparator = [](LeftType left_val, RightType right_val) { return !value_smaller(left_val, right_val); };
+      break;
+
+    default:
+      Fail(std::string("Unsupported operator for join."));
+  }
+
+  auto left_table = _left_in_table;
+  auto right_table = _right_in_table;
+
+  auto left_column_id = _left_column_id;
+  auto right_column_id = _right_column_id;
+
+  if (_mode == JoinMode::Right) {
+    // for Right Outer we swap the tables so we have the outer on the "left"
+    left_table = _right_in_table;
+    right_table = _left_in_table;
+
+    left_column_id = _right_column_id;
+    right_column_id = _left_column_id;
+  }
+
+  auto pos_list_left = std::make_shared<PosList>();
+  auto pos_list_right = std::make_shared<PosList>();
+
+  const auto is_outer_join = (_mode == JoinMode::Left || _mode == JoinMode::Right || _mode == JoinMode::Outer);
+
+  // for Full Outer, remember the matches on the right side
+  std::set<RowID> right_matches;
+
+  // Scan all chunks from left input
+  for (ChunkID chunk_id_left = ChunkID{0}; chunk_id_left < left_table->chunk_count(); ++chunk_id_left) {
+    auto column_left = left_table->get_chunk(chunk_id_left).get_column(left_column_id);
+
+    resolve_column_type<LeftType>(*column_left, [&](auto& typed_left_column) {
+      auto iterable_left = create_iterable_from_column<LeftType>(typed_left_column);
+
+      // for Outer joins, remember matches on the left side
+      std::vector<bool> left_matches;
+
+      if (is_outer_join) {
+        left_matches.resize(typed_left_column.size());
+      }
+
+      // Scan all chunks for right input
+      for (ChunkID chunk_id_right = ChunkID{0}; chunk_id_right < right_table->chunk_count(); ++chunk_id_right) {
+        auto column_right = right_table->get_chunk(chunk_id_right).get_column(right_column_id);
+
+        resolve_column_type<RightType>(*column_right, [&](auto& typed_right_column) {
+          auto iterable_right = create_iterable_from_column<RightType>(typed_right_column);
+
+          iterable_left.for_each([&](const auto& left_value) {
+            if (left_value.is_null()) return;
+
+            iterable_right.for_each([&](const auto& right_value) {
+              if (right_value.is_null()) return;
+
+              if (_comparator(left_value.value(), right_value.value())) {
+                pos_list_left->emplace_back(RowID{chunk_id_left, left_value.chunk_offset()});
+                pos_list_right->emplace_back(RowID{chunk_id_right, right_value.chunk_offset()});
+
+                if (is_outer_join) {
+                  left_matches[left_value.chunk_offset()] = true;
+                }
+
+                if (_mode == JoinMode::Outer) {
+                  right_matches.insert(RowID{chunk_id_right, right_value.chunk_offset()});
+                }
+              }
+            });
+          });
+        });
+      }
+
+      if (is_outer_join) {
+        // add unmatched rows for outer joins
```

Comment:
"for left / full outer joins"?

Reply:
all outer joins, so Left, Right and Full Outer. I thought "outer" would be a good term for it

Reply:
But here you only add the rows that are missing on one side, correct?

Reply:
for the "outer" side, yes. So for a Left Outer and Full Outer join I would add unmatched rows from the left here, and unmatched rows from the (original) right side for a Right Outer join (because for Right Outer joins we swap the tables). Does that make sense?

We are adding unmatched rows of one side, but for all three types of outer joins

Reply:
I know what you mean. My problem is that further down you "add all unmatched rows for the right side".

Leaving swapped tables aside, you are doing one part here (which is what I refer to as "left") and the other part at ` +  if (_mode == JoinMode::Outer) { `

Reply:
Now I understand. changed